// stryker/src/reporters/ProgressAppendOnlyReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var os = require("os");
var ProgressKeeper_1 = require("./ProgressKeeper");
var Timer_1 = require("../utils/Timer");
var ProgressAppendOnlyReporter = /** @class */ (function (_super) {
    tslib_1.__extends(ProgressAppendOnlyReporter, _super);
    function ProgressAppendOnlyReporter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ProgressAppendOnlyReporter.prototype.onAllMutantsMatchedWithTests = function (matchedMutants) {
        var _this = this;
        _super.prototype.onAllMutantsMatchedWithTests.call(this, matchedMutants);
        this.timer = new Timer_1.default();
        this.intervalReference = setInterval(function () { return _this.render(); }, 10000);
    };
    ProgressAppendOnlyReporter.prototype.onAllMutantsTested = function () {
        clearInterval(this.intervalReference);
    };
    ProgressAppendOnlyReporter.prototype.render = function () {
        process.stdout.write("Mutation testing " + this.procent() + " (ETC " + this.etc() + ") " +
            ("[" + this.progress.killed + " " + this.progress.killedLabel + "] ") +
            ("[" + this.progress.survived + " " + this.progress.survivedLabel + "] ") +
            ("[" + this.progress.noCoverage + " " + this.progress.noCoverageLabel + "] ") +
            ("[" + this.progress.timeout + " " + this.progress.timeoutLabel + "] ") +
            ("[" + this.progress.error + " " + this.progress.errorLabel + "]") +
            os.EOL);
    };
    ProgressAppendOnlyReporter.prototype.procent = function () {
        return Math.floor(this.progress.testedCount / this.progress.totalCount * 100) + '%';
    };
    ProgressAppendOnlyReporter.prototype.etc = function () {
        var etcSeconds = Math.floor(this.timer.elapsedSeconds() / this.progress.testedCount * (this.progress.totalCount - this.progress.testedCount));
        if (isFinite(etcSeconds)) {
            return etcSeconds + 's';
        }
        else {
            return 'n/a';
        }
    };
    return ProgressAppendOnlyReporter;
}(ProgressKeeper_1.default));
exports.default = ProgressAppendOnlyReporter;
