// stryker/src/isolated-runner/TestRunnerDecorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var events_1 = require("events");
var TestRunnerDecorator = /** @class */ (function (_super) {
    tslib_1.__extends(TestRunnerDecorator, _super);
    function TestRunnerDecorator(testRunnerProducer) {
        var _this = _super.call(this) || this;
        _this.testRunnerProducer = testRunnerProducer;
        _this.createInnerRunner();
        return _this;
    }
    TestRunnerDecorator.prototype.init = function () {
        if (this.innerRunner.init) {
            return this.innerRunner.init() || Promise.resolve();
        }
        else {
            return Promise.resolve();
        }
    };
    TestRunnerDecorator.prototype.createInnerRunner = function () {
        this.innerRunner = this.testRunnerProducer();
    };
    TestRunnerDecorator.prototype.run = function (options) {
        return this.innerRunner.run(options);
    };
    TestRunnerDecorator.prototype.dispose = function () {
        if (this.innerRunner.dispose) {
            return this.innerRunner.dispose() || Promise.resolve();
        }
        else {
            return Promise.resolve();
        }
    };
    return TestRunnerDecorator;
}(events_1.EventEmitter));
exports.default = TestRunnerDecorator;
